File: perfmonkey/R_script/parse_disk_perfmonkey.R
#!/usr/bin/env Rscript
# parse_disk_perfmonkey.R
# Converts Huawei Disk performance data to PerfMonkey format
# WITHOUT tidyverse dependency, WITH automatic SN extraction,
# WITH time sorting, and creates SEPARATE files for each disk

args <- commandArgs(trailingOnly = TRUE)

timedate_fmt <- "%Y/%m/%d %H:%M"

if (length(args) == 0) {
  stop("At least one argument must be supplied <input file> [<timedate fmt>])", call.=FALSE)
} else {
  input_file <- args[1]
  if (length(args) > 1) {
    timedate_fmt <- args[2]
  }
}

cat("\n═══════════════════════════════════════════════════════════════\n")
cat("  DISK PERFORMANCE → PERFMONKEY FORMAT CONVERTER\n")
cat("═══════════════════════════════════════════════════════════════\n\n")

label <- "RGPERF"
mpcnt <- "1"
pgcnt <- "1"
ldcnt <- "1"
alias <- "-"

# Extract Serial Number from filename
cat("Extracting Serial Number from filename...\n")
filename <- basename(input_file)
filename_clean <- gsub("\\.csv.*$", "", filename)
sn_match <- regmatches(filename_clean, regexpr("[0-9A-Z]{15,}", filename_clean))
if (length(sn_match) > 0) {
  serial <- sn_match[1]
  cat(sprintf("  ✓ Extracted Serial Number: %s\n", serial))
} else {
  serial <- "111111"
  cat("  ⚠ WARNING: Could not extract SN from filename, using default: 111111\n")
}

cat("\nReading CSV file...\n")
# Read CSV with semicolon separator
data <- read.csv(input_file, sep=";", stringsAsFactors = FALSE)
cat(sprintf("  ✓ Read %d rows\n", nrow(data)))

# Clean metric names for PerfMonkey compatibility
cat("\nCleaning metric names...\n")
data$Metric <- gsub(",", "_", data$Metric)
data$Metric <- gsub("∞", "8", data$Metric)  # infinity symbol
data$Metric <- gsub('"', "", data$Metric)   # remove quotes if any
cat("  ✓ Cleaned metric names (replaced commas and special chars)\n")

# Sort data by InstanceName and Time for proper ordering
cat("\nSorting data by instance and time...\n")
data <- data[order(data$InstanceName, data$Time), ]
cat("  ✓ Data sorted\n")

metrics <- unique(data$Metric)
instances <- unique(data$InstanceName)

cat(sprintf("\nData summary:\n"))
cat(sprintf("  • Unique instances (disks): %d\n", length(instances)))
cat(sprintf("  • Unique metrics: %d\n", length(metrics)))
cat(sprintf("  • Disk IDs: %s\n", paste(head(instances, 10), collapse=", ")))
if (length(instances) > 10) {
  cat(sprintf("    ... and %d more\n", length(instances) - 10))
}

# Process each instance (disk) separately to handle different timepoints
cat("\nProcessing each disk separately...\n")
all_instance_data <- list()

for(inst in instances) {
  cat(sprintf("\n  Processing disk: %s\n", inst))
  
  # Filter data for this instance
  inst_data <- subset(data, InstanceName == inst)
  
  # Get unique timepoints for this instance
  inst_timepoints <- unique(inst_data$Time)
  cat(sprintf("    • Timepoints: %d\n", length(inst_timepoints)))
  cat(sprintf("    • Rows: %d\n", nrow(inst_data)))
  
  # Split data by metric
  dataf_list <- split(inst_data, f=inst_data$Metric)
  
  # Extract only Time and Value columns for each metric
  dataf_list <- lapply(dataf_list, function(df) {
    df[, c("Time", "Value")]
  })
  
  # Start with first metric
  merged_df <- dataf_list[[1]]
  colnames(merged_df) <- c("Time", metrics[1])
  
  # Merge all other metrics
  if (length(metrics) > 1) {
    for(i in 2:length(metrics)) {
      temp_df <- dataf_list[[i]]
      colnames(temp_df) <- c("Time", metrics[i])
      merged_df <- merge(merged_df, temp_df, by="Time", all=TRUE)
    }
  }
  
  # Create output dataframe for this instance
  inst_output <- data.frame(
    Label = rep(label, nrow(merged_df)),
    BgnDateTime = strftime(as.POSIXct(merged_df$Time, format=timedate_fmt), 
                           format="%m/%d/%y %H:%M:%S"),
    stringsAsFactors = FALSE
  )
  inst_output$EndDateTime <- inst_output$BgnDateTime
  inst_output$Serial <- rep(serial, nrow(merged_df))
  inst_output$MpCnt <- rep(mpcnt, nrow(merged_df))
  inst_output$Rg <- rep(inst, nrow(merged_df))  # Disk ID
  inst_output$PgCnt <- rep(pgcnt, nrow(merged_df))
  inst_output$LdCnt <- rep(ldcnt, nrow(merged_df))
  inst_output$Alias <- rep(alias, nrow(merged_df))
  
  # Add metric columns
  for(metric in metrics) {
    inst_output[[metric]] <- merged_df[[metric]]
  }
  
  all_instance_data[[inst]] <- inst_output
  cat(sprintf("    ✓ Processed %d rows\n", nrow(inst_output)))
}

cat("\n═══════════════════════════════════════════════════════════════\n")
cat("  Writing separate files for each disk...\n")
cat("═══════════════════════════════════════════════════════════════\n\n")

output_files <- c()
for(inst in names(all_instance_data)) {
  inst_df <- all_instance_data[[inst]]
  
  # Add row numbers as second column (after Label)
  inst_df <- cbind(
    inst_df[, 1, drop=FALSE],  # Label column
    "#" = seq(from = 1, to = nrow(inst_df), by = 1),
    inst_df[, 2:ncol(inst_df)]  # Rest of columns
  )
  
  # Set column names
  colnames(inst_df) <- c("RGPERF", "#", "BgnDateTime", "EndDateTime", "Serial",
                          "MpCnt", "Rg", "PgCnt", "LdCnt", "Alias", metrics)
  
  # Create output filename for this disk
  # Clean disk name for filename (replace dots and slashes)
  inst_clean <- gsub("\\.", "_", inst)
  inst_clean <- gsub("/", "_", inst_clean)
  
  inst_output_file <- gsub("\\.csv.*$", "", input_file)
  inst_output_file <- paste0(inst_output_file, "_", inst_clean, "_output.csv")
  
  # Write to file WITHOUT QUOTES
  write.table(inst_df, file = inst_output_file, sep = ",",
              row.names = FALSE, col.names = TRUE,
              quote = FALSE)  # NO QUOTES!
  
  output_files <- c(output_files, inst_output_file)
  cat(sprintf("  ✓ Disk %s: %s (%d rows)\n", inst, basename(inst_output_file), nrow(inst_df)))
}

cat("\n═══════════════════════════════════════════════════════════════\n")
cat("  ✅ COMPLETED SUCCESSFULLY!\n")
cat("═══════════════════════════════════════════════════════════════\n\n")
cat(sprintf("Total files created: %d\n", length(output_files)))
cat(sprintf("Serial Number: %s\n", serial))
cat(sprintf("Metrics per file: %d\n", length(metrics)))
cat("\nFiles are ready for PerfMonkey import! 🚀\n\n")

File: perfmonkey/R_script/parse_disk_perfmonkey.py
#!/usr/bin/env python3
"""
parse_disk_perfmonkey.py - Disk performance parser for PerfMonkey format
Python replacement for parse_disk.R with memory-efficient streaming processing

Features:
- Automatic serial number extraction from filename
- Metric name cleaning (commas, infinity symbols)
- Time sorting
- Separate output files for each disk
- Memory-efficient processing for large files
- Parallel-ready (can be called multiple times)
"""

import sys
import csv
import re
from datetime import datetime
from pathlib import Path
from collections import defaultdict

# PerfMonkey format constants
LABEL = "RGPERF"
MPCNT = "1"
PGCNT = "1"
LDCNT = "1"
ALIAS = "-"
TIMEDATE_FMT = "%Y/%m/%d %H:%M"
OUTPUT_FMT = "%m/%d/%y %H:%M:%S"

def extract_serial(filename):
    """Extract serial number from filename (15+ alphanumeric chars)"""
    match = re.search(r'[0-9A-Z]{15,}', filename)
    if match:
        return match.group(0)
    return "111111"

def clean_metric_name(metric):
    """Clean metric name for PerfMonkey compatibility"""
    metric = metric.replace(',', '_')
    metric = metric.replace('∞', '8')
    metric = metric.replace('"', '')
    return metric

def parse_datetime(time_str):
    """Parse datetime from various formats"""
    # Try ISO format first (2025-09-11T02:05:00Z)
    try:
        if 'T' in time_str:
            dt = datetime.fromisoformat(time_str.replace('Z', '+00:00'))
            return dt.strftime(OUTPUT_FMT)
    except:
        pass
    
    # Try default format
    try:
        dt = datetime.strptime(time_str, TIMEDATE_FMT)
        return dt.strftime(OUTPUT_FMT)
    except:
        pass
    
    return time_str

def process_disk_file(input_file):
    """
    Process disk performance file and create separate output files for each disk
    Memory-efficient: processes data in streaming mode
    """
    input_path = Path(input_file)
    
    print("\n═══════════════════════════════════════════════════════════════")
    print("  DISK PERFORMANCE → PERFMONKEY FORMAT CONVERTER (Python)")
    print("═══════════════════════════════════════════════════════════════\n")
    
    # Extract serial number
    serial = extract_serial(input_path.name)
    print(f"Extracting Serial Number from filename...")
    print(f"  ✓ Extracted Serial Number: {serial}\n")
    
    # First pass: collect instances and metrics
    print("Reading CSV file...")
    instances = set()
    metrics_global = set()
    
    with open(input_file, 'r', encoding='utf-8') as f:
        reader = csv.DictReader(f, delimiter=';')
        for row in reader:
            instances.add(row['InstanceName'])
            metrics_global.add(clean_metric_name(row['Metric']))
    
    instances = sorted(instances)
    metrics_global = sorted(metrics_global)
    
    print(f"  ✓ Read file")
    print(f"\nData summary:")
    print(f"  • Unique instances (disks): {len(instances)}")
    print(f"  • Unique metrics: {len(metrics_global)}")
    print(f"  • Disk IDs: {', '.join(instances[:10])}")
    if len(instances) > 10:
        print(f"    ... and {len(instances) - 10} more")
    
    # Process each instance separately
    print("\nProcessing each disk separately...\n")
    output_files = []
    
    for inst in instances:
        print(f"  Processing disk: {inst}")
        
        # Collect data for this instance only
        inst_data = defaultdict(dict)  # inst_data[time][metric] = value
        timepoints = set()
        
        with open(input_file, 'r', encoding='utf-8') as f:
            reader = csv.DictReader(f, delimiter=';')
            for row in reader:
                if row['InstanceName'] == inst:
                    time = row['Time']
                    metric = clean_metric_name(row['Metric'])
                    value = row['Value']
                    inst_data[time][metric] = value
                    timepoints.add(time)
        
        # Sort timepoints
        timepoints = sorted(timepoints)
        
        print(f"    • Timepoints: {len(timepoints)}")
        print(f"    • Rows: {len(inst_data)}")
        
        # Get metrics for this instance
        metrics = sorted(set(metric for time_data in inst_data.values() for metric in time_data.keys()))
        
        # Create output filename
        inst_clean = inst.replace('.', '_').replace('/', '_')
        output_file = str(input_path).replace('.csv', f'_{inst_clean}_output.csv')
        
        # Write output file
        with open(output_file, 'w', encoding='utf-8', newline='') as f:
            writer = csv.writer(f)
            
            # Write header
            header = ['RGPERF', '#', 'BgnDateTime', 'EndDateTime', 'Serial',
                      'MpCnt', 'Rg', 'PgCnt', 'LdCnt', 'Alias'] + metrics
            writer.writerow(header)
            
            # Write data rows
            for row_num, time in enumerate(timepoints, 1):
                formatted_time = parse_datetime(time)
                row_data = [
                    LABEL,
                    row_num,
                    formatted_time,
                    formatted_time,
                    serial,
                    MPCNT,
                    inst,
                    PGCNT,
                    LDCNT,
                    ALIAS
                ]
                
                # Add metric values
                for metric in metrics:
                    row_data.append(inst_data[time].get(metric, ''))
                
                writer.writerow(row_data)
        
        output_files.append(output_file)
        print(f"    ✓ Processed {len(timepoints)} rows\n")
    
    # Summary
    print("═══════════════════════════════════════════════════════════════")
    print("  ✅ COMPLETED SUCCESSFULLY!")
    print("═══════════════════════════════════════════════════════════════\n")
    print(f"Total files created: {len(output_files)}")
    print(f"Serial Number: {serial}")
    print(f"Metrics per file: varies by disk\n")
    
    print("Output files created:")
    for output_file in output_files:
        print(f"  ✓ {Path(output_file).name}")
    
    print("\nFiles are ready for PerfMonkey import! 🚀\n")

if __name__ == '__main__':
    if len(sys.argv) < 2:
        print("Usage: python3 parse_disk_perfmonkey.py <input_file>")
        print("\nExample:")
        print("  python3 parse_disk_perfmonkey.py 2102353TJWFSP3100020.csv_disk_all")
        sys.exit(1)
    
    input_file = sys.argv[1]
    
    if not Path(input_file).exists():
        print(f"Error: File not found: {input_file}")
        sys.exit(1)
    
    process_disk_file(input_file)

